src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -50,10 +54,10 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
             //Print sources, but not for generic list (which is offline)
             if (listPackageArgs.ReportType != ReportType.Default)
             {
-                Console.WriteLine();
-                Console.WriteLine(Strings.ListPkg_SourcesUsedDescription);
-                ProjectPackagesPrintUtility.PrintSources(listPackageArgs.PackageSources);
-                Console.WriteLine();
+                reportRenderer.WriteLine();
+                reportRenderer.WriteLine(Strings.ListPkg_SourcesUsedDescription);
+                ProjectPackagesPrintUtility.PrintSources(listPackageArgs);
+                reportRenderer.WriteLine();
```

Comment:
I think that this, coupled with the fact that `JsonRenderer` has a bunch of methods that are no-ops, and the fact that there's a `if (render is JsonRenderer)` below, that there isn't a good match in abstractions.

Effectively, this `ListPackageCommandRunner` is written in a way that orders its work based on the console output, and then you're trying to bolt a json renderer on top of this.

Instead, I propose something similar to a web MVC architecture. A "controller" which gets the information for a single project (the model), and then the view (renderer) knows how to render one of these models. Something like:

```cs
renderer.Start(); // no-op for command line, for json it writes `{` and version property
foreach (var project in allProjects)
{
  // args is set up before the loop, from the command line arguments. It tells the controller
  // what properties of the model to populate, which is how it will know to be "offline only"
  // for certain text output scenarios.
  var model = await controller.GetPackagesAsync(project, args);
  renderer.Write(project);
}
renderer.End(); // json writes errors, other info?.   does console output any "wrapping up" info?
```

Honestly, the "model" could actually be the model for JSON output, and then the renderer is little more than a `JsonSerializer.Serialize(console.out, model)`, and the only place "logic" is needed is the text renderer.

This way `IReportRenderer` doesn't need a bunch of methods that are only implemented in one of the two renderers.

Reply:
Thank you for feedback. With this approach we couldn't print all the problems on top of json report, instead it would be written to the bottom of report. Is it ok?
Current template is 
```
{
 "version": 1,
  "parameters": "",
  "problems": [
    {
        "project": "src/lib/MyProjectB.csproj",
        "message": "The project `C:/Users/userA/repos/MainApp/src/lib/MyProjectB.csproj`` uses package.config for NuGet packages, while the command works only with package reference projects."
    }
  ],
  "projects": [
    {
      ...
    }
  ]
}
```

Instead, we would do:
```

{
    "version": 1,
     "parameters": "",
         "projects": [
       {
         ...
       }
     ],
     "problems": [
        {
            "project": "src/lib/MyProjectB.csproj",
            "message": "The project `C:/Users/userA/repos/MainApp/src/lib/MyProjectB.csproj`` uses package.config for NuGet packages, while the command works only with package reference projects."
        }
      ]     
}



```

Reply:
> With this approach we couldn't print all the problems on top of json report, instead it would be written to the bottom of report. Is it ok?

Two things:

1. Section 1 of the [JSON specification](https://datatracker.ietf.org/doc/html/rfc7159#section-1) says:

    >    An object is an unordered collection of zero or more name/value
    >   pairs, where a name is a string and a value is a string, number,
    >   boolean, null, object, or array.
    >
    >   An array is an ordered sequence of zero or more values.

    In other words, it is "invalid" to depend on specific order of properties in an object.  Having said that, we really, really need version to be the very first property so that tools can implement polymorphic deserialization without buffering the entire json (for when a version 2 output eventually exists).

2. The JSON renderer's implementation of `void Write(ProjectInfo)` could be just to add the object into a list, and it doesn't actually output anything until the `void End()` at the end of the loop.

The big picture I was trying to raise awareness over is overall readability and "software architecture". Let's avoid leaky abstractions. Don't get bogged down in the example comments that were designed to help understand the API design. How different renderers implement the APIs are implementation specific.

Reply:
> 
> 1. Section 1 of the [JSON specification](https://datatracker.ietf.org/doc/html/rfc7159#section-1) says:
>    > An object is an unordered collection of zero or more name/value
>    > pairs, where a name is a string and a value is a string, number,
>    > boolean, null, object, or array.
>    > An array is an ordered sequence of zero or more values.
>    
>    
>    In other words, it is "invalid" to depend on specific order of properties in an object.  Having said that, we really, really need version to be the very first property so that tools can implement polymorphic deserialization without buffering the entire json (for when a version 2 output eventually exists).

Of course, it doesn't make any difference for machine which order is objects are in json, but someone looking at the report have to scroll down all the way to find out what was the issue. Current structure allows to find out if there was issue at 1st glance and it's consistent what `npm -ls json` doing it, I feel having errors/problems at top is more user friendly.

> 2. The JSON renderer's implementation of `void Write(ProjectInfo)` could be just to add the object into a list, and it doesn't actually output anything until the `void End()` at the end of the loop.
> 

I added `void Write(ReportProject reportProject)` method into `IRenderer interface` and `Void End()` is doing the action rendering, maybe this could be not enough to address your concern.

Reply:
> but someone looking at the report have to scroll down all the way to find out what was the issue

The feature is called "machine readable".  You're right that this doesn't mean we should make it difficult for humans to read.  But customers are expected to use it in scripts or pipe to other commands, not to output to the console and look at manually.  Those scripts they create can reply the errors to their CI logs, to the console, to whereever. Alternatively, if the customer runs the command without `--format json`, we expect them to see the same errors (but in a more human friendly text output).

>  I feel having errors/problems at top is more user friendly.

Only if the output is piped to something like `more` or `less`, right?  If you just run `dotnet list pacakges --format json` and there are pages and pages of output, when the terminal stops scrolling, you're only going to see what's at the end, not at the beginning.

> I added void Write(ReportProject reportProject) method into IRenderer interface and Void End() is doing the action rendering, maybe this could be not enough to address your concern, let me know if I need to make more drastic change.

I feel that since both the text renderer and json renderer have interface methods which they ignore, it's a code smell that it's not well designed.  I get that it's quicker because you don't need to de-tangle business logic from presentation. But think about it from the point of view of someone who hasn't joined the NuGet team yet (or who doesn't review this PR). In the future they'll get some work item to work on `dotnet list package`, they'll be reading the code for the first time, and trying to understand why it's calling `renderer.WriteLine(someString)` followed by `renderer.Write(ProjectInfo)`. Why doesn't `renderer.Write(ProjectInfo` generate `someString` itself?  Another also, I know this isn't a performance hot path, but all those times where the business logic does `renderer.WriteLine(string.Format(...))`, the runtime will need to parse the format string, create a string builder, create the string, return to caller, which then gets passed to the json renderer, just to ignore it.  Not a performance hot path, but every code path is an opportunity to practise best practices.

Reply:
> I feel that since both the text renderer and json renderer have interface methods which they ignore, it's a code smell that it's not well designed. 

I agree with this. 

The abstraction is there in order to facilitate fewer code changes which isn't good design. 

Part of the reason why we've always had trouble with making changes to dotnet list package has been the lack of proper abstractions. I feel like this exacerbates that.